### DISCUSSION TITLE
Hooks into table search

### DISCUSSION THREAD

**Discussion by judgej** [filamentphp/filament#8325] [Ideas]
Just exploring an idea here, and wondered if there was already a way to do it that I haven't noticed yet, or if there is any guidance on what would need to be changed to achive it.

Basically, I'm looking for a way to hook into the table search terms, so that some terms could be pulled out completely (before being passed to the columns' search handlers) and used as a part of the global search (i.e. like a hidden filter).

An example could be a prefix like `order:123` which could be used to match on that order. Or perhaps an ISBN has been added to the search - `978-1-56619-909-4` - and I would want to recognise that, pull it out, format it for internal storage (`09781566199094`) and do a global filter on that against the main table query.

That term would then need to be removed from the search string otherwise it will result in no matches when applied to the column search methods.

From what I can tell, a search term is split into terms (separated by a space) then for each term, the term is passed to each column that is searchable. *Each term* must match *at least one* (*OR* logic) column search to return a record. By pulling the term out to handle separately, the term then can be handled like a filter where matching records must be selected for every filter that is enabled (*AND* logic).

So, is there a way to do this now? I couldn't see a way to hook in. The `InteractsWithTableQuery::applySearchConstraint()` method is on columns, not the table, so difficult to override globally (does every column type parse the whole search string individually?). So I guess I'm looking for a hook somewhere before any of the columns see the search string. It must be able to modify the search string that gets passed on to the columns, but not change how it looks on the front end.

The reason I would like this, is to provide some table filters that the system links to, but without cluttering up the table UI with too many user filters. Plus the ISBN thing is a real use-case. Someone may also be searching for `1-56619-909-3` and I would like to transform that into `09781566199094` as a filter on the main query, because they are equivalent ISBNs but we only store one normalised format.

Is this a doable thing now? If not, would it be a useful thing to offer?

------

Just as an aside, I see this in `InteractsWithTableQuery`:

```php
        if ($this->searchQuery) {
            $whereClause = $isFirst ? 'where' : 'orWhere';

            $query->{$whereClause}(
                fn ($query) => $this->evaluate($this->searchQuery, [
                    'query' => $query,
                    'search' => $search,
                ]),
            );

            $isFirst = false;

            return $query;
        }
```

I don't think there is a need to catch the "first" query differently. Laravel will do that, and is happy to accept a bunch of `orWhere()` clauses with no `where` clause. There could be some other complicating factors I'm not spotting though.

**Comment by danharrin** [comment]
`applySearchToTableQuery()` is the method to override on the list page / relation manager / Livewire component

**Reply by dvdinbox** [reply]
Can you point me to any documentation regarding overriding a method from a Resource?

I would like to have more control of the search phrase, but I can not figure out how to apply your suggestion of overriding "applySearchToTableQuery()" from within my Resource.

Example Resource: 

```
namespace App\Filament\Resources;

use App\Models\InventoryItem;
use Filament\Resources\Resource;
use Filament\Tables\Columns\TextColumn;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Builder;

class InventoryItemResource extends Resource
{
    protected static ?string $model = InventoryItem::class;

    protected static ?string $navigationIcon = 'heroicon-o-circle-stack';

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
                TextColumn::make('id')->searchable(query: function (Builder $query, string $search): Builder {
                    return $query->whereHas('product', function($q) use ($search) {
                        $q->where('title', 'like', "%$search%");
                        });
                }),               
            ]);
    }
}
```

I'm ultimately trying to circumvent the extractTableSearchWords functionality which splits the search term up if there are any spaces.

https://github.com/filamentphp/filament/blob/3.x/packages/tables/src/Concerns/CanSearchRecords.php#L110%20

cc: @judgej

**Reply by danharrin** [reply]
The method should be added to the list page class instead of the resource class

**Comment by judgej** [comment]
Thanks, I'll take a look at that.

**Comment by judgej** [comment]
Looks like I can override `getTableSearchQuery()` to filter the query string itself. Here I can pull out any matching strings that have special meaning for my custom searches, and put them aside in new property in the component. Then I can return what is left of the string to go through the normal search (global and column searches).

Then `applySearchToTableQuery()` can be extended to add in my new filters that use the saved query tokens.

**Comment by judgej** [comment]
OI noticed the original `applyColumnSearchToTableQuery()` splits the terms by a single space, so if you enter two terms and put a bunch of spaces in betweemn them, the table query gets padded out with an enormous number of where-conditions. `getTableSearchQuery()` trims the string, but I think it also needs to collapse multiple spaces, if its job is to also clean up the string somewhat. Either that, or the `explode(' ')` in `applyColumnSearchToTableQuery()` should explode on white space of any length.

Happy to do a PR, but I'm still on Filament 2. Are PRs being accepted for v2 still?

**Reply by danharrin** [reply]
I accept PRs to v2 as long as there is a corresponding one for v3 too :)

**Comment by judgej** [comment]
I see in V3 the private `getTableSearchQuery()` has gone, and replaced with a public `getTableSearch()`. It also no longer trims the query string and makes it lower case. This makes it a little harder to test, if the removal of multiple spaces is done in `applyGlobalSearchToTableQuery()` (since it is effectively an extension of the `trim()`.

So a question: is the intention of `getTableSearch()` to return the unadaulterated non-cleaned query string the user has entered, spaces and letter case preserved?

It can still be overriden to remove any special tokens from the query string (for handling elsewhere) before it gets passed on, but I guess no other cleaning should be done on the string?

**Reply by danharrin** [reply]
It should preserve capitalisation, since it's used for the search indicator in the toolbar, so shown to the user as-is. Trimming probably doesn't hurt though 👍

**Reply by judgej** [reply]
I see. So `testy    test    ` in the query box shows this in the address bar:

    tableSearchQuery=testy++++++test++++

So back to my original requirement, if I override `getTableSearch()` to remove some words for use in a custom filter, then those words no longer appear in the address bar? I don't think it works that way in v2, where the address bar reflects the value of `$tableSearchQuery` only.

**Reply by danharrin** [reply]
It does work like that in V2, it's the behaviour of Livewire's query string

**Reply by judgej** [reply]
Then I've misunderstood. I thought you were saying the output of `getTableSearch()` went to the address bar. It's `$tableSearch` that is bound to the address bar. `getTableSearch()` is what the table global and table column filters see. Which is good.

**Comment by judgej** [comment]
Having trouble creating a PR - it keeps getting closed - so I'm going to dump some details here while I still have them. This is for V3.

I have a simple table on the User model, showing `name` and `email`. I dump the database query that `applyGlobalSearchToTableQuery()` generates before any changes.

Witn no search terms:

    select * from `users` for 

With two search terms URL shows `tableSearch=name+Example`:

    select * from `users` where (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?)

Now, if it put some extra spaces in the search terms it provides this query parameter `tableSearch=name+++++++Example` which is fine, but the query is now:

    select * from `users` where (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?) and (`name` like ? or `email` like ?)

My PR fixes this so the URL remains unchanged, but the resulting query is back to the shorter version.

There is a new protected method `searchWords()` that does the word splitting. This serves multiple purposes:

* Testable on its own.
* Can be used to parse out the words when overridding `getTableSearch()` as a convenience. The `getTableSearch()` still has access to the original query string provided by the user, so functionality has not changed for filters where that is important.

~~The [PR I tried creating is here](https://github.com/filamentphp/filament/pull/8375). I went through the PR process again and could not see any settings where permissions can be changed.~~

PR here: https://github.com/filamentphp/filament/pull/8392

**Reply by danharrin** [reply]
In your fork, you need to create a new branch from 3.x, otherwise maintainers cant edit it
